ref(ci): Use a single CI workflow for tests

gustavovalverde opened the pull request:
## Motivation

This is an initial implementation to solve the re-building of our Docker image, which is being built multiple times by our Github Actions, causing elevated costs across multiple services.

Fixes #7816

### PR Author Checklist

#### Check before marking the PR as ready for review:
  - [ ] Will the PR name make sense to users?
  - [ ] Does the PR have a priority label?
  - [ ] Have you added or updated tests?
  - [ ] Is the documentation up to date?

##### For significant changes:
  - [ ] Is there a summary in the CHANGELOG?
  - [ ] Can these changes be split into multiple PRs?

_If a checkbox isn't relevant to the PR, mark it as done._

### Complex Code or Requirements

- This required creating an extra workflow `ci-tests.yml` with its respectives `.patch.yml` and `.patch-external.yml` to initialize the unit tests and integration tests from a single place, effectively using a single Docker image build process.
- This must be admin merged, and all open PRs must be rebased afterwards.

## Solution

- Create a `Run tests` [starter workflow](https://docs.github.com/en/actions/using-workflows/reusing-workflows#reusable-workflows-and-starter-workflows)
- Convert the `Integration Tests on GCP` and `Docker Unit Tests` to [callable workflows](https://docs.github.com/en/actions/using-workflows/reusing-workflows#reusable-workflows-and-starter-workflows)
- Merge both of the `Integration Tests on GCP` and `Docker Unit Tests` patches files into a single patch file for `Run tests`

### Testing

- Tests should run on this PR successfully


### Reviewer Checklist

Check before approving the PR:
  - [ ] Does the PR scope match the ticket?
  - [ ] Are there enough tests to make sure it works? Do the tests cover the PR motivation?
  - [ ] Are all the PR blockers dealt with?
        PR blockers can be dealt with in new tickets or PRs.

_And check the PR Author checklist is complete._

## Follow Up Work

This must be followed by https://github.com/ZcashFoundation/zebra/pull/8374 to effectively reduce our costs.

gustavovalverde committed:
ref(ci): Use a single CI workflow for tests

This is an initial implementation to solve the re-building of our Docker image, which is being built multiple times by our Github Actions.

This is meant to fix -> devops: Build CI Docker runs twice for every PR #7816

gustavovalverde committed:
fix(patch): use correct name reference

gustavovalverde committed:
chore: reduce diff with correct sentence

gustavovalverde committed:
Merge branch 'main' into fix-duplicate-builds

upbqdn commented on .github/workflows/ci-tests.yml:
```diff
@@ -0,0 +1,148 @@
+      # workflow definitions
+      - "docker/**"
+      - ".dockerignore"
+      - ".github/workflows/ci-tests.yml"
```
Should we also add the following workflows?

```suggestion
      - ".github/workflows/ci-tests.yml"
      - ".github/workflows/sub-ci-unit-tests-docker.yml"
      - ".github/workflows/sub-ci-integration-tests-gcp.yml"
```

upbqdn commented on .github/workflows/ci-tests.yml:
```diff
@@ -0,0 +1,148 @@
+        default: false
+
+  pull_request:
+    # Skip PRs where Rust code and dependencies aren't modified.
```
###### Optional Suggestion

```suggestion
    # Run only on PRs that modify Rust code or dependencies.
```

upbqdn commented on .github/workflows/ci-tests.yml:
```diff
@@ -0,0 +1,148 @@
+      - ".github/workflows/sub-build-docker-image.yml"
+
+  push:
+    # Skip main branch updates where Rust code and dependencies aren't modified.
```
###### Optional Suggestion

```suggestion
    # Run only on main branch updates that modify Rust code or dependencies.
```

upbqdn commented on .github/workflows/ci-tests.patch-external.yml:
```diff
@@ -1,115 +1,139 @@
-  get-available-disks:
-    name: Check if cached state disks exist for Mainnet / Check if cached state disks exist
+  build:
+    name: Build CI Docker / Build images
```
Should we change the name to match the job name in `ci-tests.yml`?

```suggestion
    name: Build CI Docker / Build CI image
```

upbqdn reviewed (commented):
This PR removes workflow patches for unit tests, but it doesn't do so for integration tests. Why is that?

upbqdn commented on .github/workflows/ci-tests.patch.yml:
```diff
@@ -1,45 +1,98 @@
-  get-available-disks:
-    name: Check if cached state disks exist for Mainnet / Check if cached state disks exist
+  build:
+    name: Build CI Docker / Build images
```
Related to a previous comment.

```suggestion
    name: Build CI Docker / Build CI image
```

gustavovalverde committed:
Apply suggestions from code review

Co-authored-by: Marek <mail@marek.onl>

gustavovalverde commented:
> This PR removes workflow patches for unit tests, but it doesn't do so for integration tests. Why is that?

The changelog is confusing, but I removed the unit tests patches, and renamed and reused the same patch files for integration tests, those are the new `ci-tests.patch.yml` and `ci-tests.patch-external.yml`. This was mainly to reduce the diff.

upbqdn commented on .github/workflows/ci-tests.patch-external.yml:
```diff
@@ -1,115 +1,139 @@
+
+
+  ####
+  ## The following jobs are related to sub-ci-intgration-tests-gcp.yml
```
```suggestion
  ## The following jobs are related to sub-ci-integration-tests-gcp.yml
```

upbqdn commented on .github/workflows/ci-tests.patch.yml:
```diff
@@ -1,45 +1,98 @@
+
+
+  ####
+  ## The following jobs are related to sub-ci-intgration-tests-gcp.yml
```
```suggestion
  ## The following jobs are related to sub-ci-integration-tests-gcp.yml
```

gustavovalverde committed:
Apply suggestions from code review

Co-authored-by: Marek <mail@marek.onl>

gustavovalverde commented:
@upbqdn after merging your last suggestions, this will require re-approval. Thanks!

gustavovalverde committed:
fix(rules): revert job names, to avoid creating an extra rule

gustavovalverde commented:
@upbqdn Oops! I had to revert one of the suggestions as I had forgotten the reasoning behind it.

The job is called `Build images` as the specific workflow that's being (`./.github/workflows/sub-build-docker-image.yml`) called is used for **CI**, **CD** and **Release** images.

Also, changing the job name here would then require 2 merge rules instead of 1:

Now:
- `Build CI Docker / Build images` and `Build CD Docker / Build images` (1 rule)
After:
- `Build CI Docker / Build CI image` (1 rule)
- `Build CD Docker / Build CD image` (1 rule)

This would also be a breaking change, requiring other PRs to be updated. I'd prefer to avoid that in this refactor.

upbqdn commented on .github/workflows/ci-tests.patch.yml:
```diff
@@ -1,45 +1,98 @@
-  get-available-disks:
-    name: Check if cached state disks exist for Mainnet / Check if cached state disks exist
+  build:
+    name: Build CI Docker / Build CI image
```
I guess we should change this one then as well.

```suggestion
    name: Build CI Docker / Build images
```

gustavovalverde committed:
Update .github/workflows/ci-tests.patch.yml

Co-authored-by: Marek <mail@marek.onl>
